storage/src/desktop/storage_reference_desktop.cc

```diff
@@ -279,18 +277,96 @@ Future<size_t> StorageReferenceInternal::GetBytes(void* buffer,
+    PrepareRequest(request, storageUri_.AsHttpUrl().c_str(), rest::util::kGet);
+    GetBytesResponse* response =
+        new GetBytesResponse(buffer, buffer_size, handle, future_api);
+    RestCall(request, request->notifier(), response, handle.get(), listener,
+             controller_out);
+    return response;
+  }};
+  SendRequestWithRetry(kStorageReferenceFnGetBytesInternal, send_request_funct,
+                       handle, storage_->max_download_retry_time());
+  return GetBytesLastResult();
+}
 
-  GetBytesResponse* response =
-      new GetBytesResponse(buffer, buffer_size, handle, future_api);
+// Sends a rest request, and creates a separate thread to retry failures.
+void StorageReferenceInternal::SendRequestWithRetry(
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct, SafeFutureHandle<size_t> final_handle,
+    double max_retry_time_seconds) {
+  BlockingResponse* first_response = send_request_funct();
+  std::thread async_retry_thread(
+      &StorageReferenceInternal::AsyncSendRequestWithRetry, this,
+      internal_function_reference, send_request_funct, final_handle,
+      first_response, max_retry_time_seconds);
+  async_retry_thread.detach();
+}
+
+const int kInitialSleepTimeMillis = 1000;
+const int kMaxSleepTimeMillis = 30000;
+
+// In a separate thread, repeatedly send Rest requests until one succeeds or a
+// maximum amount of time has passed.
+void StorageReferenceInternal::AsyncSendRequestWithRetry(
+    StorageReferenceFn internal_function_reference,
+    SendRequestFunct send_request_funct, SafeFutureHandle<size_t> final_handle,
+    BlockingResponse* response, double max_retry_time_seconds) {
+  auto* future_api = future();
+  firebase::FutureBase internal_future;
+  auto end_time = std::chrono::steady_clock::now() +
+                  std::chrono::duration<double>(max_retry_time_seconds);
+  auto current_sleep_time = std::chrono::milliseconds(kInitialSleepTimeMillis);
+  while (true) {
+    internal_future = future_api->LastResult(internal_function_reference);
+    // Wait for completion, then check status and error.
+    while (internal_future.status() == firebase::kFutureStatusPending) {
+      std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+    // For any request that succeeds or fails in a non-retryable way, don't
+    // bother retrying.
+    if (internal_future.status() != firebase::kFutureStatusComplete ||
+        !IsRetryableFailure(response->status())) {
+      break;
+    }
+    // Interrupt the loop if the retry deadline has been reached
+    auto current_time = std::chrono::steady_clock::now();
+    if (current_time + current_sleep_time > end_time) {
+      break;
+    }
+    // Sleep for an exponentially increasing duration, then retry the request.
+    std::this_thread::sleep_for(current_sleep_time);
+    if (current_sleep_time < std::chrono::milliseconds(kMaxSleepTimeMillis)) {
+      current_sleep_time *= 2;
```

Comment:
Probably best to cap it at kMaxSleepTimeMillis rather than *=2 if it's below that; otherwise, the max it will reach right now is 32000 rather than the max of 30000.

Reply:
Done, I now set to max_sleep_time if it exceeds max_sleep_time.